src/core/search/compressed_list.cc

```diff
@@ -0,0 +1,213 @@
+  // so we need to encode both differences diff1 and diff2
+  DCHECK_GT(bound.value, value);
+  DCHECK_LE(bound.prev_value, value);
+
+  // Compute and encode new diff1 and diff2 into buf1 and buf2 respectivaly
+  array<uint8_t, kMaxBufferSize> buf1, buf2;
+  auto diff1_span = WriteVarLen(value - bound.prev_value, absl::MakeSpan(buf1));
+  auto diff2_span = WriteVarLen(bound.value - value, absl::MakeSpan(buf2));
+
+  // Extend the location where diff0 is stored with optional zeros before overwriting it
+  ptrdiff_t diff_offset = bound.diff_span.data() - diffs_.data();
+  size_t required_len = diff1_span.size() + diff2_span.size();
+  DCHECK_LE(bound.diff_span.size(), required_len);  // It can't shrink for sure
+  diffs_.insert(diffs_.begin() + diff_offset, required_len - bound.diff_span.size(), 0u);
+
+  // Now overwrite diff0 and 0s with the two new differences
+  copy(diff1_span.begin(), diff1_span.end(), diffs_.begin() + diff_offset);
+  copy(diff2_span.begin(), diff2_span.end(), diffs_.begin() + diff_offset + diff1_span.size());
+}
+
+// Remove has linear complexity. It tries to find the element V and its neighbors A and B,
+// which are encoded as diff1 = V - A and diff2 = B - V. Adjacently stored diff1 and diff2
+// need to be replaced with diff3 = diff1 + diff2s
+void CompressedList::Remove(IntType value) {
+  auto bound = LowerBound(value);
+
+  // Nothing was read or the element was not found
+  if (bound.diff_span.empty() || bound.value != value)
+    return;
+
+  // We're removing below unconditionally
+  size_--;
+
+  // Calculate offset where values diff is stored and determine diffs tail
+  ptrdiff_t diff_offset = bound.diff_span.data() - diffs_.data();
+  auto diffs_tail = absl::MakeSpan(diffs_).subspan(diff_offset + bound.diff_span.size());
+
+  // If it's stored at the end, simply truncate it away
+  if (diffs_tail.empty()) {
+    diffs_.resize(diffs_.size() - bound.diff_span.size());
+    return;
+  }
+
+  // Now the list certainly contains a succeeding element B > V and possibly A < V (or 0)
+  // Read diff2 and calculate diff3 = diff1 + diff2
+  auto [diff2, diff2_read] = ReadVarLen(diffs_tail);
+  IntType diff3 = (bound.value - bound.prev_value) + diff2;
+
+  // Encode diff3
+  array<uint8_t, kMaxBufferSize> buf;
+  auto diff3_buf = WriteVarLen(diff3, absl::MakeSpan(buf));
+
+  // Shrink vector before overwriting
+  DCHECK_LE(diff3_buf.size(), diff2_read + bound.diff_span.size());
+  size_t to_remove = diff2_read + bound.diff_span.size() - diff3_buf.size();
+  diffs_.erase(diffs_.begin() + diff_offset, diffs_.begin() + diff_offset + to_remove);
+
+  // Overwrite diff1/diff2 with new diff3
+  copy(diff3_buf.begin(), diff3_buf.end(), diffs_.begin() + diff_offset);
+}
+
+size_t CompressedList::Size() const {
+  return size_;
+}
+
+size_t CompressedList::ByteSize() const {
+  return diffs_.size();
+}
+
+absl::Span<uint8_t> CompressedList::WriteVarLen(IntType value, absl::Span<uint8_t> buf) {
+  buf[0] = (value & 0b11111) << 3;
```

Comment:
I see that you're using a constant 3-bit header here. That's an interesting tradeoff. I guess that it's both simpler to implement, and also saves some bytes for large numbers. However if you expect many diffs in the range of `[2^5, 2^7)` and `[2^13, 2^15)` that would be less efficient.
Your guess is better than mine, so whatever you choose, it's your call.

However, I do think it would be more readable to add some
```cpp
constexpr unsigned kHeaderSize = 3;
```
And use that, as well as `8 - kHeaderSize`.
Perhaps also `0b11111` could be written as `0b1'11'11`
Also, side note, it's more common to use the leftmost bits for the non-data bits, but your call :)

Reply:
1. Yes, but we'll have to tune the decoding later either way, I'd better not spend many more iterations on compression if it will change in the future, what's here is comparable to what we had
2. Yes, will add to constexpr
3. Yes, storing non data in prefix is better. Will swap them